### DISCUSSION TITLE
Velocity not working like paper Velocity

### DISCUSSION THREAD

**Discussion by TropicalShadow** [Minestom/Minestom#1456] [Help] [answered]
I am trying to create a Launchpad, and shoot the player x block in the direction of the player
```kt
val direction = player.position.direction().mul(10.0)
player.velocity = direction
```
But it doesn't seem to be moving the player correctly. it shoves the player a block or so, 
even though the Direction == 9.633643460563258 -2.6387305317633762 0.47959863999653934

This is a paper version, and it works as intended.
```java
player.setVelocity(location.getDirection().multiply(launch))
```

>This could just be my lack of understanding Velocity and Direction... but from what I can tell it doesn't function correctly.
Minestom: 1.19.2

**Comment by Moulberry** [comment, marked as answer]
In Minestom, velocity is sent to the player using the "Set Entity Velocity" packet:
https://wiki.vg/Protocol#Set_Entity_Velocity

Because this packet uses shorts, it is limited to +/- 4 blocks of velocity.
If you want to change the velocity above that, you have to use the "Explosion" packet:
https://wiki.vg/Protocol#Explosion

Paper clamps the velocity to +/- 3.9, whereas Minestom lets the short value overflow.
This means if you try to set a velocity of +9.63, the actual value the client receives is +1.63.
This is definitely an issue with Minestom that I have brought up before, but I don't think a fix was decided on.

If you want a quick solution, just modify your code to clamp the velocity to +/- 3.9 like paper does automatically

**Reply by TropicalShadow** [reply]
After reading your suggestion I opened up Paper Source code and found where they handled Velocity and made this from it.
```kt

fun Player.addVelocity(x: Double, y: Double, z: Double) {
    val e: Double = MathUtils.clamp(x, -3.9, 3.9)
    val f: Double = MathUtils.clamp(y, -3.9, 3.9)
    val g: Double = MathUtils.clamp(z, -3.9, 3.9)
    this.sendPacketToViewersAndSelf(EntityVelocityPacket(this.entityId,
        (e * 8000.0).toInt().toShort(),
        (f * 8000.0).toInt().toShort(),
        (g * 8000.0).toInt().toShort()
    ))
}
```
[wiki.vg](https://wiki.vg/Protocol#Set_Entity_Velocity) says (Velocity is believed to be in units of 1/8000 of a block per server tick (50ms)). 
so I presume that's where the Magical 8000 number comes from.
> This works... although I don't know what will happen due to not actually changing the player's velocity variable server side. (if there is one)

**Reply by Obyvante** [reply]
@TropicalShadow Thank you for your solution but I have noticed strange behaviors with the packet-based velocity solution. For example, immediate teleportation to velocity motion.

@Moulberry Does paper use explode packet for this or is there any solution to have a "proper" velocity system?

**Reply by Obyvante** [reply]
Entity velocity is not working like a player. This is interesting. Maybe due to something in velocityTick?

**Reply by Obyvante** [reply]
So I have decided not to touch the current velocity system. Instead, I will just clamp for short;,

```java
    public EntityVelocityPacket(int entityId, Point velocity) {
        this(entityId,
                (short) MathUtils.clamp(velocity.x(), Short.MIN_VALUE, Short.MAX_VALUE),
                (short) MathUtils.clamp(velocity.y(), Short.MIN_VALUE, Short.MAX_VALUE),
                (short) MathUtils.clamp(velocity.z(), Short.MIN_VALUE, Short.MAX_VALUE));
    }
```
